### DISCUSSION TITLE
List subfolders and files of a folder (with SAF), Android 13+

### DISCUSSION THREAD

**Discussion by MaTrics** [apache/cordova#494] [General]
Hey together!

I'm already researching and trying for 1,5 days now but i don't get the point.
My goal: My app let's the user select a folder (external storage, so he can upload files there). Then in my app i want to get the subfolders and the files (jpg's) inside of them, just reading them, no manipulation.

I now that cordova-plugin-file doesn't help anymore since we bound to SAF.

I can do the folder selection with cordova-plugin-saf-mediastore. Let's say, the user selects the folder "books", so i get the URI "content://com.android.externalstorage.documents/tree/primary%3Abooks"
I can't call saf-plugin's getFilename, this get's my a Java Exception about unsupported URI type.
Also window.resolveLocalFileSystemURL() doesn't work, what i didn't expect.

Is that possible at all in JS/Cordova without switching to Java coding?

**Comment by breautek** [comment]
> Is that possible at all in JS/Cordova without switching to Java coding?

Short of an existing plugin that already does these tasks, no I don't think so.

The file plugin does have limited support for `content://` paths but I don't think it's possible to obtain the real filename. The filename I believe is derived from the URI which obviously for content:// paths, doesn't contain the underlying filename or display name.

It's weird that your SAF plugin isn't accepting `content://com.android.externalstorage.documents/tree/primary%3Abooks`, perhaps it's not properly decoding the URI?

**Reply by MaTrics** [reply]
Thanks! Hm ok, so what i really want is: in a user selected folder, get the sub folders and inside oft them load the image files. So i wouldn't need to know the rea filename. I only have to get the references to them to load them in my HTML.
What i don't understand: could i achieve this with the cordova-plugin-file (under Android API 34)?

**Reply by breautek** [reply]
> could i achieve this with the cordova-plugin-file (under Android API 34)?

Kind of, but not completely.

The main piece that is missing for your use case I think is the open document intent implementation. That's the system that provides the user the document view so that they can select a file. The file plugin can resolve `content://` paths and read from them, but using the file plugin itself, there is no way to obtain `content://` paths.

It also doesn't request permissions. When you use that system dialog so that the user can select content, the system will grant your app temporary read access to that file which expires when your app's activity closes. This means that the `content://` paths are only temporarily accessible so they cannot be persistently stored. So if you need permanent access to that content, that content needs to be copied to your app's storage data.

There is also a bug that prevents `content://` paths to produce DOM-usable urls so that content resources can be used in `<img>` tags or other HTML media elements. There is a PR proposal to fix this issue via https://github.com/apache/cordova-plugin-file/pull/629 though.

I've been looking into making other Apache plugins like the camera and media-capture plugins return `content://` paths instead of a `file://` path which has a number of advantages. So I've been doing quite a bit testing on using `content://` paths in the file plugin and most of it works today, assuming that:

1. You're just trying to do read operations on the content
2. You've obtained the `content://` path freshly in a way that the OS will give you read access to that content
3. Using the FileReader APIs should work, or using the `copyTo` API to copy to app's local data storage should also work.

So in otherwords today, if your SAF plugin can provide you a `content://` path with read access granted, the file plugin should be able to be used to read or copy that data. The only thing that won't work is trying to produce a DOM-usable URL which requires the https://github.com/apache/cordova-plugin-file/pull/629 fix.

**Reply by MaTrics** [reply]
Thanks so much for your detailed answer. In my investigations i didn't manage to feed the URI provided by cordova-plugin-saf-mediastore into the file-plugin. I think the big problems are indeed the permissions. I'll find a totally other workaround. Thank you and best regards!

**Reply by breautek** [reply]
On permissions it's important to understand that under Android's SAF framework, apps don't have broad access to external storage, they have temporary granular access instead and most cases it's read only. Write access depends on the underlying content provider.

So taking a primitive example of a social media app, that allows you to upload a profile picture. It may access an existing file on external storage, gain temporary read access to that file. But you don't want to store and use the provided `content://` path because that path will expire once the app is closed. The app will want to retain access to it long-term. Therefore the app should make a copy of that resource while it has the read access and use it's own local copy instead of the original `content://` resource.

**Reply by MaTrics** [reply]
Thanks for explanation. Ok, but so i could then copy the whole folder structure at that point into tha app's file ordner. But anyway, i don't manage to get access at all, so i will go another way now and try to download it all as zip from a server into the app's files folder.
